### DISCUSSION TITLE
pag能不能应用到容器的背景上

### DISCUSSION THREAD

**Discussion by xujixiao** [Tencent/libpag#1823] [Q&A] [answered]
pag能不能应用到容器的背景上，可以放置组件在动画内部，不是分层重叠的方式，pag在下，组件在上，例如一个面版是一个pag动画做的收起和展开效果，但是面版内有内容，希望可以pag可以像容器组件一样承载内部的非容器view在安卓中

**Comment by kevingpqi123** [comment]
指的是 pag 内部嵌入系统的 UI 控件吗？建议直接描述清楚你要实现的功能，然后我们看下这里的最优实现方案，有可能和你现在想的实现方案不一样

**Reply by xujixiao** [reply]
就是我现在有个礼物面版，面版中有很多礼物，这个礼物面版的展开和收起是一个pag的动画效果，目前我实现方案是用的pagimageview执行的这个收起和展开的pag动画，面版中的礼物用的普通的安卓原生的控件，盖在pagImageView上面，导致我很多兼容性的问题，间距不好控制，所以希望能把这个pagImageview的动画控件，当成ViewGroup之类的容器控件，能把礼物放在里面，这样好控制间距

**Reply by xujixiao** [reply]
就是希望pagImageView之类的控件内部能嵌入，类似于textview imageview之类的android 系统控件，而不是盖在上面

**Reply by kevingpqi123** [reply]
如果是这类需求，面板中的礼物可以不通过原生控件来实现，通过 pag 来实现就可以了哈，让设计师将礼物设计进入 pag 文件中，通过 getLayersUnderPoint 接口检测那些空间被点击，通过 setMatrix 可以进一步调整位置

pag 内部渲染树和系统原生控件是两套不同的渲染逻辑，不能将系统控件嵌入到 pag 的渲染容器中，但 PAGView 或 PAGImageView 本质上就是一个 View，可以在此基础上做上层封装

**Reply by xujixiao** [reply]
我们的礼物面版内的原生控件还有很多复杂的逻辑，所以不太适合做成pag动画的形式，那样过于复杂了，pag动画在不同的分辨率的设备上，感觉不好控制兼容性，pag动画中存在一个icon，这个icon 在宽屏和分辨率比较高的设备看着比较模糊，不能自适应屏幕分辨率，原生控件在高分辨率的设备上看着比较小，但是清晰，这个问题应该怎么解决比较好呢

**Reply by kevingpqi123** [reply, marked as answer]
pag 解决的是 AE 动效还原的问题，设计师在 AE 中设计的动效可以快速上线于各平台端，如果你这边有比较复杂的逻辑，通过原生方案来做是更好的，pag 存在的目的是解决动效的问题，而不是替换原生方案
如果你想各个平台的清晰度较好，那就采用最大屏幕清晰度的素材，就比如展示一张图片一样，如果你想各个设备看着都很清晰，首先要有最大屏幕清晰度的素材，然后在较低屏幕分辨率机型上就是向小缩放，清晰度也可以得到保障
